feat: `no-control-regex` support `v` flag

ota-meshi opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[X] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

Refs #17223

This PR modifies the `no-control-regex` rule and adds support for regexp `v` flag.

#### Is there anything you'd like reviewers to focus on?

<!-- markdownlint-disable-file MD004 -->

nzakas reviewed (changes requested):
Code LGTM. Can we add in some more tests that use regex literals, too?

ota-meshi commented:
Thank you for reviewing this PR. I added two testcases. Could you please check again?

nzakas reviewed (approved):
LGTM, thanks! Would like @mdjermanovic to verify before merging.

mdjermanovic reviewed (commented):
Some weird behavior with `"uv"`:

```js
/* eslint no-control-regex: ["error"] */

/\x11/; // error: Unexpected control character(s) in regular expression: \x11.

RegExp("", "uv"); // error: Unexpected control character(s) in regular expression: \x11.

```

I'm guessing this happens because in this case `validatePattern` throws before `onPatternEnter` is called, so moving `this._controlChars = [];` into `collectControlChars` might help (though I'm not sure why it wasn't already there, i.e. is there a reason why `onPatternEnter` was used for this?).

ota-meshi commented:
Thank you for finding the bug!

I added `this._controlChars = []` to `collectControlChars`.
However, I left the `onPatternEnter`. I added a comment as to why. Could you please check that?

```js
        /*
         * `RegExpValidator` may parse the pattern twice in one `validatePattern`.
         * So `this._controlChars` should be cleared here as well.
         *
         * For example, the `/(?<a>\x1f)/` regex will parse the pattern twice.
         * This is based on the content described in Annex B.
         * If the regex contains a `GroupName` and the `u` flag is not used, `ParseText` will be called twice.
         * See https://tc39.es/ecma262/2023/multipage/additional-ecmascript-features-for-web-browsers.html#sec-parsepattern-annexb
         */
```

mdjermanovic commented:
Thanks for the explanation, very useful! I hadn't thought of that.

mdjermanovic reviewed (approved):
LGTM, thanks!
